store/src/java/com/zimbra/cs/ml/classifier/ClassifierRegistry.java

```diff
@@ -0,0 +1,197 @@
+            Classifier<?> decoded = decode(encoded);
+            String id = decoded.getId();
+            ClassifierInfo info = infoMap.get(id);
+            if (info == null) {
+                ZimbraLog.ml.warn("known classifier %s (id=%s) has no entry in ML backend, deleting", decoded.getLabel(), id);
+                deRegister(encoded);
+            } else {
+                ZimbraLog.ml.debug("loaded classifier label=%s id=%s", decoded.getLabel(), id);
+                infoMap.remove(id);
+                idMap.put(id, decoded);
+                labelMap.put(decoded.getLabel(), decoded);
+                decoded.setClassifierInfo(info);
+            }
+        }
+        if (!infoMap.isEmpty()) {
+            for (String cId: infoMap.keySet()) {
+                ZimbraLog.ml.warn("classifier id=%s has no corresponding entry in %s", cId, this.getClass().getSimpleName());
+            }
+        }
+        loaded = true;
+    }
+
+    public Map<String, Classifier<?>> getAllClassifiers() throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        return idMap;
+    }
+
+    public boolean labelExists(String label) throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        return labelMap.containsKey(label);
+    }
+
+    /**
+     * return known encoded classifiers
+     */
+    protected abstract String[] load() throws ServiceException;
+
+    /**
+     * Persist the encoded classifier string
+     */
+    protected abstract void save(String encodedClassifier) throws ServiceException;
+
+    /**
+     * Register a classifier
+     */
+    public void register(Classifier<?> classifier) throws ServiceException {
+        if (!loaded) {
+            loadKnownClassifiers();
+        }
+        if (labelExists(classifier.getLabel())) {
+            throw ServiceException.FAILURE("label '" + classifier.getLabel() + "' already exists", null);
+        }
+        idMap.put(classifier.getId(), classifier);
+        labelMap.put(classifier.getLabel(), classifier);
+        save(encode(classifier));
+    }
+
+    protected String encode(Classifier<?> classifier) {
+        Map<String, Object> map = new HashMap<>();
+        map.put(KEY_ID, classifier.getId());
+        map.put(KEY_TYPE, classifier.getType().name());
+        map.put(KEY_LABEL, classifier.getLabel());
+        map.put(KEY_FEATURES, classifier.getFeatureSet().getAllFeatureSpecs().stream().map(feature -> ((FeatureSpec<?>) feature).encode()).collect(Collectors.toList()));
+        map.put(KEY_DESCRIPTION, classifier.getDescription());
+        return BEncoding.encode(map);
+    }
+
+    protected Classifier<?> decode(String encoded) throws ServiceException {
+        Map<String, Object> map;
+        try {
+            map = BEncoding.decode(encoded);
+        } catch (BEncodingException e) {
+            throw ServiceException.FAILURE("unable to decode classifier", null);
```

Comment:
Worth logging the value of `encoded`?

Reply:
The encoded values can be pretty long, but it's probably worth logging in error cases